python/tank/descriptor/descriptor.py

```diff
@@ -162,10 +141,47 @@ class Descriptor(object):
 
     (APP, FRAMEWORK, ENGINE, CONFIG, CORE, INSTALLED_CONFIG) = range(6)
 
+    _factory = {}
+
+    @classmethod
+    def register_descriptor_factory(cls, descriptor_type, subclass):
```

Comment:
in order to realise this pattern, the factory is now split into smaller chunks. We are making things more complicated to avoid local imports. I am not sure i like it.